build_runner/lib/src/build/asset_graph/nodes.dart

```diff
@@ -0,0 +1,265 @@
+      _invalidatedSortedFileIdsByPackage = true;
+      _sortedFileIdsByPackage = null;
+    }
+
+    return updatedNode;
+  }
+
+  /// Adds [node] to the graph if it doesn't exist.
+  ///
+  /// Throws a [StateError] if it already exists in the graph.
+  ///
+  /// Returns the updated node: if it replaces an [AssetNode.missingSource] then
+  /// `outputs` and `primaryOutputs` are copied to it from that.
+  AssetNode add(AssetNode node) {
+    if (node.isFile && _sortedFileIdsByPackage != null) {
+      _invalidatedSortedFileIdsByPackage = true;
+      _sortedFileIdsByPackage = null;
+    }
+    final existing = get(node.id);
+    if (existing != null) {
+      if (existing.type == NodeType.missingSource) {
+        _nodes.remove(existing.id);
+        node = node.rebuild((b) {
+          b.primaryOutputs.addAll(existing.primaryOutputs);
+        });
+      } else {
+        throw StateError(
+          'Tried to add node ${node.id} to the asset graph but it already '
+          'exists.',
+        );
+      }
+    }
+    _nodes[node.id] = node;
+    if (node.inputs?.isNotEmpty ?? false) {
+      _outputs = null;
+    }
+
+    return node;
+  }
+
+  /// Removes [id] from the graph, if present.
+  void remove(AssetId id) {
+    final removed = _nodes.remove(id);
+    if (removed != null && removed.isFile && _sortedFileIdsByPackage != null) {
+      _invalidatedSortedFileIdsByPackage = true;
+      _sortedFileIdsByPackage = null;
+    }
+  }
+
+  /// All nodes in the graph.
+  Iterable<AssetNode> get allNodes => _nodes.values;
+
+  /// All IDs in `package` that refer to files and match.
+  ///
+  /// A node refers to a file if [AssetNode.isFile].
+  ///
+  /// If [glob] is passed, return only IDs for which the glob matches the path.
+  Iterable<AssetId> packageFileIds(String package, {Glob? glob}) {
+    if (_invalidatedSortedFileIdsByPackage) {
+      throw StateError('Sorted file IDs by package were invalidated.');
+    }
+    _sortedFileIdsByPackage ??= _computeSortedFileIdsByPackage();
+    final list = _sortedFileIdsByPackage![package];
+    if (list == null) return const [];
+    if (glob == null) return list;
+
+    final simplePrefix = _simpleGlobPrefix(glob);
+    Iterable<AssetId> result = list;
+    if (simplePrefix.isNotEmpty) {
+      result = list.filterToPrefix(simplePrefix);
+    }
+
+    return result.where((id) => glob.matches(id.path));
+  }
+
+  /// The prefix of the [glob] pattern up to but not including the first special
+  /// character.
+  String _simpleGlobPrefix(Glob glob) {
+    final pattern = glob.pattern;
+    for (var i = 0; i != pattern.length; ++i) {
+      final codeUnit = pattern.codeUnitAt(i);
+
+      /// Code units signalling the start of non-literal content in a glob
+      /// pattern, see: https://pub.dev/packages/glob
+      ///
+      /// Checked in nodes_test.dart:
+      ///
+      /// `expect(r'*?[\{'.codeUnits, [42, 63, 91, 92, 123])`
+      switch (codeUnit) {
```

Comment:
Potentially you could add

```dart
const int $STAR = 42;
const int $QUESTION = 63;
const int $OPEN_SQUARE_BRACKET = 91;
const int $BACKSLASH = 92;
const int $OPEN_CURLY_BRACKET = 123;
```

(constants copied from the sdk @ pkg/_fe_analyzer_shared/lib/src/scanner/characters.dart) --- which could then make the switch more readable.